core/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -418,25 +422,35 @@ public static long nodeLongValue(Object node) {
     /**
      * This method is very lenient, use {@link #nodeBooleanValue} instead.
      */
-    public static boolean lenientNodeBooleanValue(Object node, boolean defaultValue) {
+    public static boolean lenientNodeBooleanValue(Object node, String name, boolean defaultValue) {
         if (node == null) {
             return defaultValue;
         }
-        return lenientNodeBooleanValue(node);
+        return lenientNodeBooleanValue(node, name);
     }
 
     /**
      * This method is very lenient, use {@link #nodeBooleanValue} instead.
      */
-    public static boolean lenientNodeBooleanValue(Object node) {
+    public static boolean lenientNodeBooleanValue(Object node, String name) {
+        boolean interpretedAsLenient = false;
+        boolean booleanValue;
+
         if (node instanceof Boolean) {
-            return (Boolean) node;
+            booleanValue = (Boolean) node;
+        } else if (node instanceof Number) {
+            interpretedAsLenient = true;
+            booleanValue = ((Number) node).intValue() != 0;
+        } else {
+            String value = node.toString();
+            booleanValue = !(value.equals("false") || value.equals("0") || value.equals("off"));
```

Comment:
You are right. This is the original behavior and I just moved code around. I have really mixed feelings about fixing it now because it could lead to bugs in client applications that are very hard to find. Despite this I am still slightly in favor of fixing it too because this problem would appear anyway at latest in 6.0. However, I will fix it in a separate commit after merging this PR.